{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Taxonomy ID: 907340\n",
      "Genetic Code ID: 1\n",
      "Genetic Code Name: Standard\n",
      "\n",
      "Codon Table Retrieved:\n",
      "Table 1 Standard, SGC0\n",
      "\n",
      "  |  T      |  C      |  A      |  G      |\n",
      "--+---------+---------+---------+---------+--\n",
      "T | TTT F   | TCT S   | TAT Y   | TGT C   | T\n",
      "T | TTC F   | TCC S   | TAC Y   | TGC C   | C\n",
      "T | TTA L   | TCA S   | TAA Stop| TGA Stop| A\n",
      "T | TTG L(s)| TCG S   | TAG Stop| TGG W   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "C | CTT L   | CCT P   | CAT H   | CGT R   | T\n",
      "C | CTC L   | CCC P   | CAC H   | CGC R   | C\n",
      "C | CTA L   | CCA P   | CAA Q   | CGA R   | A\n",
      "C | CTG L(s)| CCG P   | CAG Q   | CGG R   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "A | ATT I   | ACT T   | AAT N   | AGT S   | T\n",
      "A | ATC I   | ACC T   | AAC N   | AGC S   | C\n",
      "A | ATA I   | ACA T   | AAA K   | AGA R   | A\n",
      "A | ATG M(s)| ACG T   | AAG K   | AGG R   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "G | GTT V   | GCT A   | GAT D   | GGT G   | T\n",
      "G | GTC V   | GCC A   | GAC D   | GGC G   | C\n",
      "G | GTA V   | GCA A   | GAA E   | GGA G   | A\n",
      "G | GTG V   | GCG A   | GAG E   | GGG G   | G\n",
      "--+---------+---------+---------+---------+--\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Data import CodonTable\n",
    "\n",
    "Entrez.email = \"qamilmirza@berkeley.edu\"  # Required for NCBI queries\n",
    "\n",
    "# Search for the species in the NCBI taxonomy database\n",
    "species_name = \"Hansenula saturnus\"  # Change to any species of interest\n",
    "handle = Entrez.esearch(db=\"taxonomy\", term=species_name)\n",
    "record = Entrez.read(handle)\n",
    "handle.close()\n",
    "\n",
    "if record[\"IdList\"]:\n",
    "    taxon_id = record[\"IdList\"][0]\n",
    "    print(f\"Taxonomy ID: {taxon_id}\")\n",
    "\n",
    "    # Fetch taxonomy details to obtain the genetic code ID\n",
    "    handle = Entrez.efetch(db=\"taxonomy\", id=taxon_id, retmode=\"xml\")\n",
    "    records = Entrez.read(handle)\n",
    "    handle.close()\n",
    "\n",
    "    if records:\n",
    "        genetic_code = records[0].get(\"GeneticCode\", {})\n",
    "        gc_id = genetic_code.get(\"GCId\", \"N/A\")\n",
    "        gc_name = genetic_code.get(\"GCName\", \"N/A\")\n",
    "        print(f\"Genetic Code ID: {gc_id}\")\n",
    "        print(f\"Genetic Code Name: {gc_name}\")\n",
    "\n",
    "        # Print the associated codon table\n",
    "        if gc_id.isdigit():\n",
    "            gc_id = int(gc_id)\n",
    "            if gc_id in CodonTable.unambiguous_dna_by_id:\n",
    "                codon_table = CodonTable.unambiguous_dna_by_id[gc_id]\n",
    "                print(\"\\nCodon Table Retrieved:\")\n",
    "                print(codon_table)\n",
    "            else:\n",
    "                print(f\"Codon table for Genetic Code ID {gc_id} is not available in Biopython.\")\n",
    "        else:\n",
    "            print(\"Invalid Genetic Code ID.\")\n",
    "    else:\n",
    "        print(\"No records found.\")\n",
    "else:\n",
    "    print(\"Species not found.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Universal translation: MKTW\n",
      "Yeast mitochondrial translation: MKTW\n"
     ]
    }
   ],
   "source": [
    "# Advanced Project Code: Species-Dependent Translator\n",
    "from Bio.Data import CodonTable\n",
    "\n",
    "def translate_sequence(seq, species=\"universal\"):\n",
    "    \"\"\"\n",
    "    Translate a DNA sequence using the codon table appropriate for the given species.\n",
    "    \n",
    "    Parameters:\n",
    "        seq (str): The DNA sequence (should be in 5'-to-3' orientation).\n",
    "        species (str): Either \"universal\" or a string indicating another species/organellar code.\n",
    "                       For this example, if species is \"yeast_mito\", we use table 3.\n",
    "    \n",
    "    Returns:\n",
    "        str: The translated protein sequence.\n",
    "    \"\"\"\n",
    "    # Choose the appropriate codon table.\n",
    "    # (This mapping can be expanded or automated by querying NCBI taxonomy data.)\n",
    "    if species.lower() in [\"saccharomyces cerevisiae mitochondria\", \"yeast_mito\", \"yeast mitochondria\"]:\n",
    "        table_id = 3  # Yeast mitochondrial code (as defined by NCBI)\n",
    "    else:\n",
    "        table_id = 1  # Universal code\n",
    "    \n",
    "    codon_table = CodonTable.unambiguous_dna_by_id[table_id]\n",
    "    \n",
    "    protein = \"\"\n",
    "    # Process the sequence in windows of three.\n",
    "    for i in range(0, len(seq), 3):\n",
    "        codon = seq[i:i+3]\n",
    "        # Ignore incomplete codons at the end.\n",
    "        if len(codon) != 3:\n",
    "            continue\n",
    "        # Check if the codon is a stop codon.\n",
    "        if codon in codon_table.stop_codons:\n",
    "            protein += \"*\"\n",
    "        else:\n",
    "            # Look up the amino acid. If the codon is not found, use 'X' for unknown.\n",
    "            aa = codon_table.forward_table.get(codon, \"X\")\n",
    "            protein += aa\n",
    "    return protein\n",
    "\n",
    "# Example usage:\n",
    "dna_seq = \"ATGAAGACCTGG\"  # Example sequence; adjust as needed.\n",
    "protein_universal = translate_sequence(dna_seq, species=\"universal\")\n",
    "protein_yeast = translate_sequence(dna_seq, species=\"yeast_mito\")\n",
    "\n",
    "print(\"Universal translation:\", protein_universal)\n",
    "print(\"Yeast mitochondrial translation:\", protein_yeast)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Universal translation: MIVLGK*\n",
      "Yeast mitochondrial translation: MMVLGKW\n"
     ]
    }
   ],
   "source": [
    "dna_seq = \"ATGATAGTATTAGGTAAGTGA\"\n",
    "protein_universal = translate_sequence(dna_seq, species=\"universal\")\n",
    "protein_yeast = translate_sequence(dna_seq, species=\"yeast_mito\")\n",
    "\n",
    "print(\"Universal translation:\", protein_universal)\n",
    "print(\"Yeast mitochondrial translation:\", protein_yeast)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "from Bio.Data import CodonTable\n",
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "from io import StringIO\n",
    "from urllib.parse import quote \n",
    "\n",
    "def get_species_codon_table(species_name):\n",
    "    \"\"\"\n",
    "    Input: species_name (str) - the name of the species to search for\n",
    "    Output: table_id (int) - the genetic code table ID for the species\n",
    "    \"\"\"\n",
    "    encoded_species = quote(species_name)\n",
    "    search_url = f\"https://www.ebi.ac.uk/ena/taxonomy/rest/scientific-name/{encoded_species}\"\n",
    "    \n",
    "    try:\n",
    "        response = requests.get(search_url)\n",
    "        response.raise_for_status()\n",
    "        data = response.json()\n",
    "        print(data)\n",
    "        \n",
    "        if isinstance(data, list) and len(data) > 0:\n",
    "            taxon_info = data[0]\n",
    "            return int(taxon_info.get(\"geneticCode\", 1))  # Ensure integer\n",
    "        else:\n",
    "            return None\n",
    "    except requests.RequestException as e:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ðŸ” Search Query: Saccharomyces cerevisiae[Organism] AND ACT1[Gene] AND biomol_mrna[PROP] AND srcdb_refseq[PROP]\n",
      "âœ… Found IDs: ['296144393']\n",
      "ðŸ“¥ Fetching sequence for ID: 296144393...\n",
      "ðŸ“ƒ FASTA Header: >NM_001179927.1 Saccharomyces cerevisiae S288C actin (ACT1), partial mRNA\n",
      "NM_001179927.1: AUGGAUUCUGAGGUUGCUGCUUUGGUUAUUGAUAACGGUUCUGGUAUGUGUAAAGCCGGUUUUGCCGGUGACGACGCUCCUCGUGCUGUCUUCCCAUCUAUCGUCGGUAGACCAAGACACCAAGGUAUCAUGGUCGGUAUGGGUCAAAAAGACUCCUACGUUGGUGAUGAAGCUCAAUCCAAGAGAGGUAUCUUGACUUUACGUUACCCAAUUGAACACGGUAUUGUCACCAACUGGGACGAUAUGGAAAAGAUCUGGCAUCAUACCUUCUACAACGAAUUGAGAGUUGCCCCAGAAGAACACCCUGUUCUUUUGACUGAAGCUCCAAUGAACCCUAAAUCAAACAGAGAAAAGAUGACUCAAAUUAUGUUUGAAACUUUCAACGUUCCAGCCUUCUACGUUUCCAUCCAAGCCGUUUUGUCCUUGUACUCUUCCGGUAGAACUACUGGUAUUGUUUUGGAUUCCGGUGAUGGUGUUACUCACGUCGUUCCAAUUUACGCUGGUUUCUCUCUACCUCACGCCAUUUUGAGAAUCGAUUUGGCCGGUAGAGAUUUGACUGACUACUUGAUGAAGAUCUUGAGUGAACGUGGUUACUCUUUCUCCACCACUGCUGAAAGAGAAAUUGUCCGUGACAUCAAGGAAAAACUAUGUUACGUCGCCUUGGACUUCGAACAAGAAAUGCAAACCGCUGCUCAAUCUUCUUCAAUUGAAAAAUCCUACGAACUUCCAGAUGGUCAAGUCAUCACUAUUGGUAACGAAAGAUUCAGAGCCCCAGAAGCUUUGUUCCAUCCUUCUGUUUUGGGUUUGGAAUCUGCCGGUAUUGACCAAACUACUUACAACUCCAUCAUGAAGUGUGAUGUCGAUGUCCGUAAGGAAUUAUACGGUAACAUCGUUAUGUCCGGUGGUACCACCAUGUUCCCAGGUAUUGCCGAAAGAAUGCAAAAGGAAAUCACCGCUUUGGCUCCAUCUUCCAUGAAGGUCAAGAUCAUUGCUCCUCCAGAAAGAAAGUACUCCGUCUGGAUUGGUGGUUCUAUCUUGGCUUCUUUGACUACCUUCCAACAAAUGUGGAUCUCAAAACAAGAAUACGACGAAAGUGGUCCAUCUAUCGUUCACCACAAGUGUUUCUAA\n",
      "1128\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez, SeqIO\n",
    "\n",
    "def get_mrna_sequence(species_name, gene_name, email=\"qamilmirza@berkeley.edu\"):\n",
    "    \"\"\"\n",
    "    Fetch mRNA sequences from NCBI for a given species and gene name.\n",
    "\n",
    "    Parameters:\n",
    "        species_name (str): Name of the species (e.g., \"Homo sapiens\")\n",
    "        gene_name (str): Name of the gene (e.g., \"INS\")\n",
    "        email (str): Your email for NCBI Entrez (required)\n",
    "\n",
    "    Returns:\n",
    "        dict: {accession_number: mRNA_sequence (with T->U conversion)}\n",
    "    \"\"\"\n",
    "    Entrez.email = email  # Set email for NCBI Entrez API\n",
    "    \n",
    "    try:\n",
    "        # New query: using property filters for mRNA and RefSeq records\n",
    "        search_query = (f\"{species_name}[Organism] AND {gene_name}[Gene] \"\n",
    "                        \"AND biomol_mrna[PROP] AND srcdb_refseq[PROP]\")\n",
    "        print(f\"ðŸ” Search Query: {search_query}\")  # Debug: show search query\n",
    "\n",
    "        # Increase retmax if needed; here we limit to 5 records\n",
    "        handle = Entrez.esearch(db=\"nucleotide\", term=search_query, retmax=5)\n",
    "        record = Entrez.read(handle)\n",
    "        handle.close()\n",
    "        \n",
    "        print(f\"âœ… Found IDs: {record['IdList']}\")  # Debug: show list of IDs\n",
    "\n",
    "        if not record[\"IdList\"]:\n",
    "            print(\"âŒ No mRNA sequences found. Try a different query or check gene name.\")\n",
    "            return {}\n",
    "\n",
    "        mrna_sequences = {}\n",
    "\n",
    "        for gene_id in record[\"IdList\"]:\n",
    "            print(f\"ðŸ“¥ Fetching sequence for ID: {gene_id}...\")\n",
    "            fetch_handle = Entrez.efetch(db=\"nucleotide\", id=gene_id, rettype=\"fasta\", retmode=\"text\")\n",
    "            fasta_data = fetch_handle.read()\n",
    "            fetch_handle.close()\n",
    "\n",
    "            fasta_lines = fasta_data.strip().split(\"\\n\")\n",
    "            # The header is the first line; remove the initial '>' and extract accession (first token)\n",
    "            header = fasta_lines[0].split()[0][1:]\n",
    "            sequence = \"\".join(fasta_lines[1:])  # Join all sequence lines together\n",
    "            \n",
    "            # Convert DNA to RNA (T -> U)\n",
    "            rna_sequence = sequence.replace(\"T\", \"U\").replace(\"t\", \"u\")\n",
    "            mrna_sequences[header] = rna_sequence\n",
    "            \n",
    "            print(f\"ðŸ“ƒ FASTA Header: {fasta_lines[0]}\")\n",
    "        \n",
    "        return mrna_sequences  # Return dictionary {Accession ID: mRNA sequence}\n",
    "    \n",
    "    except Exception as e:\n",
    "        print(f\"ðŸš¨ Error: {e}\")\n",
    "        return {}\n",
    "\n",
    "\n",
    "species = \"Saccharomyces cerevisiae\"\n",
    "gene = \"ACT1\"\n",
    "mrna_seqs = get_mrna_sequence(species, gene)\n",
    "for acc, seq in mrna_seqs.items():\n",
    "    print(f\"{acc}: {seq}\")\n",
    "    print(len(seq))  # print the first 60 bases for brevity\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetching mRNA sequence for:\n",
      "Species: Saccharomyces cerevisiae, Gene: CYC1\n",
      "\n",
      "Using refined query:\n",
      "Saccharomyces cerevisiae[Organism] AND CYC1[Gene] AND refseq_mrna[Filter]\n",
      "\n",
      "Raw esearch response (refseq_mrna filter):\n",
      "{'Count': '0', 'RetMax': '0', 'RetStart': '0', 'IdList': [], 'TranslationSet': [{'From': 'Saccharomyces cerevisiae[Organism]', 'To': '\"Saccharomyces cerevisiae\"[Organism]'}], 'QueryTranslation': 'Saccharomyces cerevisiae[Organism] AND CYC1[Gene] AND refseq_mrna[Filter]', 'ErrorList': {'PhraseNotFound': ['refseq_mrna[Filter]'], 'FieldNotFound': []}, 'WarningList': {'OutputMessage': ['No items found.'], 'QuotedPhraseNotFound': [], 'PhraseIgnored': []}}\n",
      "\n",
      "No records found with refseq_mrna filter.\n",
      "Using fallback query:\n",
      "Saccharomyces cerevisiae[Organism] AND CYC1[Gene] AND mRNA[Filter]\n",
      "Error: Search Backend failed: \t\t\n",
      "\n",
      "No sequence retrieved.\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "Entrez.email = \"qamilmirza@berkeley.edu\"\n",
    "\n",
    "def get_mrna_sequence_with_debug(species_name, gene_name):\n",
    "    \"\"\"\n",
    "    Fetches the mRNA sequence for a given species and gene.\n",
    "    Prints out the raw responses from the Entrez queries for debugging.\n",
    "    \n",
    "    Parameters:\n",
    "      species_name (str): e.g., \"Homo sapiens\" or \"Saccharomyces cerevisiae\"\n",
    "      gene_name (str): e.g., \"INS\" or \"CYC1\"\n",
    "      \n",
    "    Returns:\n",
    "      A single mRNA sequence (string) with T's replaced by U's,\n",
    "      or an empty string if no sequence is found.\n",
    "    \"\"\"\n",
    "    try:\n",
    "        # --- First Query: Using the refined query with refseq_mrna filter ---\n",
    "        search_query = f\"{species_name}[Organism] AND {gene_name}[Gene] AND refseq_mrna[Filter]\"\n",
    "        print(\"Using refined query:\")\n",
    "        print(search_query)\n",
    "        \n",
    "        handle = Entrez.esearch(db=\"nucleotide\", term=search_query, retmax=5)\n",
    "        record = Entrez.read(handle)\n",
    "        handle.close()\n",
    "        \n",
    "        print(\"\\nRaw esearch response (refseq_mrna filter):\")\n",
    "        print(record)\n",
    "        \n",
    "        # If no records are found, fall back to a generic query using mRNA filter\n",
    "        if not record[\"IdList\"]:\n",
    "            search_query = f\"{species_name}[Organism] AND {gene_name}[Gene] AND mRNA[Filter]\"\n",
    "            print(\"\\nNo records found with refseq_mrna filter.\")\n",
    "            print(\"Using fallback query:\")\n",
    "            print(search_query)\n",
    "            \n",
    "            handle = Entrez.esearch(db=\"nucleotide\", term=search_query, retmax=5)\n",
    "            record = Entrez.read(handle)\n",
    "            handle.close()\n",
    "            \n",
    "            print(\"\\nRaw esearch response (fallback mRNA filter):\")\n",
    "            print(record)\n",
    "            \n",
    "            if not record[\"IdList\"]:\n",
    "                print(\"\\nNo mRNA records found with either query.\")\n",
    "                return \"\"\n",
    "        \n",
    "        # Select the first gene ID (assumed to be the canonical transcript)\n",
    "        gene_id = record[\"IdList\"][0]\n",
    "        print(\"\\nSelected Gene ID:\")\n",
    "        print(gene_id)\n",
    "        \n",
    "        # --- Fetch the FASTA sequence for the selected record ---\n",
    "        fetch_handle = Entrez.efetch(db=\"nucleotide\", id=gene_id, rettype=\"fasta\", retmode=\"text\")\n",
    "        fasta_data = fetch_handle.read()\n",
    "        fetch_handle.close()\n",
    "        \n",
    "        print(\"\\nRaw FASTA data retrieved:\")\n",
    "        print(fasta_data)\n",
    "        \n",
    "        # Process FASTA data: remove header and join sequence lines\n",
    "        fasta_lines = fasta_data.strip().split(\"\\n\")\n",
    "        sequence_only = ''.join(fasta_lines[1:])\n",
    "        mrna_sequence = sequence_only.replace(\" \", \"\").replace(\"T\", \"U\").replace(\"t\", \"u\")\n",
    "        \n",
    "        return mrna_sequence\n",
    "    except Exception as e:\n",
    "        print(\"Error:\", e)\n",
    "        return \"\"\n",
    "\n",
    "# Example usage in your Jupyter Notebook:\n",
    "species_name = \"Saccharomyces cerevisiae\"\n",
    "gene_name = \"CYC1\"\n",
    "\n",
    "print(\"Fetching mRNA sequence for:\")\n",
    "print(f\"Species: {species_name}, Gene: {gene_name}\\n\")\n",
    "\n",
    "mrna_sequence = get_mrna_sequence_with_debug(species_name, gene_name)\n",
    "\n",
    "if mrna_sequence:\n",
    "    print(\"\\nFinal Processed mRNA Sequence (first 100 nucleotides):\")\n",
    "    print(mrna_sequence[:100])\n",
    "    print(\"\\nTotal Sequence Length:\", len(mrna_sequence), \"nucleotides\")\n",
    "else:\n",
    "    print(\"No sequence retrieved.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "3.11.9",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
