{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Taxonomy ID: 907340\n",
      "Genetic Code ID: 1\n",
      "Genetic Code Name: Standard\n",
      "\n",
      "Codon Table Retrieved:\n",
      "Table 1 Standard, SGC0\n",
      "\n",
      "  |  T      |  C      |  A      |  G      |\n",
      "--+---------+---------+---------+---------+--\n",
      "T | TTT F   | TCT S   | TAT Y   | TGT C   | T\n",
      "T | TTC F   | TCC S   | TAC Y   | TGC C   | C\n",
      "T | TTA L   | TCA S   | TAA Stop| TGA Stop| A\n",
      "T | TTG L(s)| TCG S   | TAG Stop| TGG W   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "C | CTT L   | CCT P   | CAT H   | CGT R   | T\n",
      "C | CTC L   | CCC P   | CAC H   | CGC R   | C\n",
      "C | CTA L   | CCA P   | CAA Q   | CGA R   | A\n",
      "C | CTG L(s)| CCG P   | CAG Q   | CGG R   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "A | ATT I   | ACT T   | AAT N   | AGT S   | T\n",
      "A | ATC I   | ACC T   | AAC N   | AGC S   | C\n",
      "A | ATA I   | ACA T   | AAA K   | AGA R   | A\n",
      "A | ATG M(s)| ACG T   | AAG K   | AGG R   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "G | GTT V   | GCT A   | GAT D   | GGT G   | T\n",
      "G | GTC V   | GCC A   | GAC D   | GGC G   | C\n",
      "G | GTA V   | GCA A   | GAA E   | GGA G   | A\n",
      "G | GTG V   | GCG A   | GAG E   | GGG G   | G\n",
      "--+---------+---------+---------+---------+--\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Data import CodonTable\n",
    "\n",
    "Entrez.email = \"qamilmirza@berkeley.edu\"  # Required for NCBI queries\n",
    "\n",
    "# Search for the species in the NCBI taxonomy database\n",
    "species_name = \"Hansenula saturnus\"  # Change to any species of interest\n",
    "handle = Entrez.esearch(db=\"taxonomy\", term=species_name)\n",
    "record = Entrez.read(handle)\n",
    "handle.close()\n",
    "\n",
    "if record[\"IdList\"]:\n",
    "    taxon_id = record[\"IdList\"][0]\n",
    "    print(f\"Taxonomy ID: {taxon_id}\")\n",
    "\n",
    "    # Fetch taxonomy details to obtain the genetic code ID\n",
    "    handle = Entrez.efetch(db=\"taxonomy\", id=taxon_id, retmode=\"xml\")\n",
    "    records = Entrez.read(handle)\n",
    "    handle.close()\n",
    "\n",
    "    if records:\n",
    "        genetic_code = records[0].get(\"GeneticCode\", {})\n",
    "        gc_id = genetic_code.get(\"GCId\", \"N/A\")\n",
    "        gc_name = genetic_code.get(\"GCName\", \"N/A\")\n",
    "        print(f\"Genetic Code ID: {gc_id}\")\n",
    "        print(f\"Genetic Code Name: {gc_name}\")\n",
    "\n",
    "        # Print the associated codon table\n",
    "        if gc_id.isdigit():\n",
    "            gc_id = int(gc_id)\n",
    "            if gc_id in CodonTable.unambiguous_dna_by_id:\n",
    "                codon_table = CodonTable.unambiguous_dna_by_id[gc_id]\n",
    "                print(\"\\nCodon Table Retrieved:\")\n",
    "                print(codon_table)\n",
    "            else:\n",
    "                print(f\"Codon table for Genetic Code ID {gc_id} is not available in Biopython.\")\n",
    "        else:\n",
    "            print(\"Invalid Genetic Code ID.\")\n",
    "    else:\n",
    "        print(\"No records found.\")\n",
    "else:\n",
    "    print(\"Species not found.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Universal translation: MKTW\n",
      "Yeast mitochondrial translation: MKTW\n"
     ]
    }
   ],
   "source": [
    "# Advanced Project Code: Species-Dependent Translator\n",
    "from Bio.Data import CodonTable\n",
    "\n",
    "def translate_sequence(seq, species=\"universal\"):\n",
    "    \"\"\"\n",
    "    Translate a DNA sequence using the codon table appropriate for the given species.\n",
    "    \n",
    "    Parameters:\n",
    "        seq (str): The DNA sequence (should be in 5'-to-3' orientation).\n",
    "        species (str): Either \"universal\" or a string indicating another species/organellar code.\n",
    "                       For this example, if species is \"yeast_mito\", we use table 3.\n",
    "    \n",
    "    Returns:\n",
    "        str: The translated protein sequence.\n",
    "    \"\"\"\n",
    "    # Choose the appropriate codon table.\n",
    "    # (This mapping can be expanded or automated by querying NCBI taxonomy data.)\n",
    "    if species.lower() in [\"saccharomyces cerevisiae mitochondria\", \"yeast_mito\", \"yeast mitochondria\"]:\n",
    "        table_id = 3  # Yeast mitochondrial code (as defined by NCBI)\n",
    "    else:\n",
    "        table_id = 1  # Universal code\n",
    "    \n",
    "    codon_table = CodonTable.unambiguous_dna_by_id[table_id]\n",
    "    \n",
    "    protein = \"\"\n",
    "    # Process the sequence in windows of three.\n",
    "    for i in range(0, len(seq), 3):\n",
    "        codon = seq[i:i+3]\n",
    "        # Ignore incomplete codons at the end.\n",
    "        if len(codon) != 3:\n",
    "            continue\n",
    "        # Check if the codon is a stop codon.\n",
    "        if codon in codon_table.stop_codons:\n",
    "            protein += \"*\"\n",
    "        else:\n",
    "            # Look up the amino acid. If the codon is not found, use 'X' for unknown.\n",
    "            aa = codon_table.forward_table.get(codon, \"X\")\n",
    "            protein += aa\n",
    "    return protein\n",
    "\n",
    "# Example usage:\n",
    "dna_seq = \"ATGAAGACCTGG\"  # Example sequence; adjust as needed.\n",
    "protein_universal = translate_sequence(dna_seq, species=\"universal\")\n",
    "protein_yeast = translate_sequence(dna_seq, species=\"yeast_mito\")\n",
    "\n",
    "print(\"Universal translation:\", protein_universal)\n",
    "print(\"Yeast mitochondrial translation:\", protein_yeast)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Universal translation: MIVLGK*\n",
      "Yeast mitochondrial translation: MMVLGKW\n"
     ]
    }
   ],
   "source": [
    "dna_seq = \"ATGATAGTATTAGGTAAGTGA\"\n",
    "protein_universal = translate_sequence(dna_seq, species=\"universal\")\n",
    "protein_yeast = translate_sequence(dna_seq, species=\"yeast_mito\")\n",
    "\n",
    "print(\"Universal translation:\", protein_universal)\n",
    "print(\"Yeast mitochondrial translation:\", protein_yeast)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "from Bio.Data import CodonTable\n",
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "from io import StringIO\n",
    "from urllib.parse import quote \n",
    "\n",
    "def get_species_codon_table(species_name):\n",
    "    \"\"\"\n",
    "    Input: species_name (str) - the name of the species to search for\n",
    "    Output: table_id (int) - the genetic code table ID for the species\n",
    "    \"\"\"\n",
    "    encoded_species = quote(species_name)\n",
    "    search_url = f\"https://www.ebi.ac.uk/ena/taxonomy/rest/scientific-name/{encoded_species}\"\n",
    "    \n",
    "    try:\n",
    "        response = requests.get(search_url)\n",
    "        response.raise_for_status()\n",
    "        data = response.json()\n",
    "        print(data)\n",
    "        \n",
    "        if isinstance(data, list) and len(data) > 0:\n",
    "            taxon_info = data[0]\n",
    "            return int(taxon_info.get(\"geneticCode\", 1))  # Ensure integer\n",
    "        else:\n",
    "            return None\n",
    "    except requests.RequestException as e:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ðŸ” Search Query: Saccharomyces cerevisiae[Organism] AND ACT1[Gene] AND biomol_mrna[PROP] AND srcdb_refseq[PROP]\n",
      "âœ… Found IDs: ['296144393']\n",
      "ðŸ“¥ Fetching sequence for ID: 296144393...\n",
      "ðŸ“ƒ FASTA Header: >NM_001179927.1 Saccharomyces cerevisiae S288C actin (ACT1), partial mRNA\n",
      "NM_001179927.1: AUGGAUUCUGAGGUUGCUGCUUUGGUUAUUGAUAACGGUUCUGGUAUGUGUAAAGCCGGUUUUGCCGGUGACGACGCUCCUCGUGCUGUCUUCCCAUCUAUCGUCGGUAGACCAAGACACCAAGGUAUCAUGGUCGGUAUGGGUCAAAAAGACUCCUACGUUGGUGAUGAAGCUCAAUCCAAGAGAGGUAUCUUGACUUUACGUUACCCAAUUGAACACGGUAUUGUCACCAACUGGGACGAUAUGGAAAAGAUCUGGCAUCAUACCUUCUACAACGAAUUGAGAGUUGCCCCAGAAGAACACCCUGUUCUUUUGACUGAAGCUCCAAUGAACCCUAAAUCAAACAGAGAAAAGAUGACUCAAAUUAUGUUUGAAACUUUCAACGUUCCAGCCUUCUACGUUUCCAUCCAAGCCGUUUUGUCCUUGUACUCUUCCGGUAGAACUACUGGUAUUGUUUUGGAUUCCGGUGAUGGUGUUACUCACGUCGUUCCAAUUUACGCUGGUUUCUCUCUACCUCACGCCAUUUUGAGAAUCGAUUUGGCCGGUAGAGAUUUGACUGACUACUUGAUGAAGAUCUUGAGUGAACGUGGUUACUCUUUCUCCACCACUGCUGAAAGAGAAAUUGUCCGUGACAUCAAGGAAAAACUAUGUUACGUCGCCUUGGACUUCGAACAAGAAAUGCAAACCGCUGCUCAAUCUUCUUCAAUUGAAAAAUCCUACGAACUUCCAGAUGGUCAAGUCAUCACUAUUGGUAACGAAAGAUUCAGAGCCCCAGAAGCUUUGUUCCAUCCUUCUGUUUUGGGUUUGGAAUCUGCCGGUAUUGACCAAACUACUUACAACUCCAUCAUGAAGUGUGAUGUCGAUGUCCGUAAGGAAUUAUACGGUAACAUCGUUAUGUCCGGUGGUACCACCAUGUUCCCAGGUAUUGCCGAAAGAAUGCAAAAGGAAAUCACCGCUUUGGCUCCAUCUUCCAUGAAGGUCAAGAUCAUUGCUCCUCCAGAAAGAAAGUACUCCGUCUGGAUUGGUGGUUCUAUCUUGGCUUCUUUGACUACCUUCCAACAAAUGUGGAUCUCAAAACAAGAAUACGACGAAAGUGGUCCAUCUAUCGUUCACCACAAGUGUUUCUAA\n",
      "1128\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez, SeqIO\n",
    "\n",
    "def get_mrna_sequence(species_name, gene_name, email=\"qamilmirza@berkeley.edu\"):\n",
    "    \"\"\"\n",
    "    Fetch mRNA sequences from NCBI for a given species and gene name.\n",
    "\n",
    "    Parameters:\n",
    "        species_name (str): Name of the species (e.g., \"Homo sapiens\")\n",
    "        gene_name (str): Name of the gene (e.g., \"INS\")\n",
    "        email (str): Your email for NCBI Entrez (required)\n",
    "\n",
    "    Returns:\n",
    "        dict: {accession_number: mRNA_sequence (with T->U conversion)}\n",
    "    \"\"\"\n",
    "    Entrez.email = email  # Set email for NCBI Entrez API\n",
    "    \n",
    "    try:\n",
    "        # New query: using property filters for mRNA and RefSeq records\n",
    "        search_query = (f\"{species_name}[Organism] AND {gene_name}[Gene] \"\n",
    "                        \"AND biomol_mrna[PROP] AND srcdb_refseq[PROP]\")\n",
    "        print(f\"ðŸ” Search Query: {search_query}\")  # Debug: show search query\n",
    "\n",
    "        # Increase retmax if needed; here we limit to 5 records\n",
    "        handle = Entrez.esearch(db=\"nucleotide\", term=search_query, retmax=5)\n",
    "        record = Entrez.read(handle)\n",
    "        handle.close()\n",
    "        \n",
    "        print(f\"âœ… Found IDs: {record['IdList']}\")  # Debug: show list of IDs\n",
    "\n",
    "        if not record[\"IdList\"]:\n",
    "            print(\"âŒ No mRNA sequences found. Try a different query or check gene name.\")\n",
    "            return {}\n",
    "\n",
    "        mrna_sequences = {}\n",
    "\n",
    "        for gene_id in record[\"IdList\"]:\n",
    "            print(f\"ðŸ“¥ Fetching sequence for ID: {gene_id}...\")\n",
    "            fetch_handle = Entrez.efetch(db=\"nucleotide\", id=gene_id, rettype=\"fasta\", retmode=\"text\")\n",
    "            fasta_data = fetch_handle.read()\n",
    "            fetch_handle.close()\n",
    "\n",
    "            fasta_lines = fasta_data.strip().split(\"\\n\")\n",
    "            # The header is the first line; remove the initial '>' and extract accession (first token)\n",
    "            header = fasta_lines[0].split()[0][1:]\n",
    "            sequence = \"\".join(fasta_lines[1:])  # Join all sequence lines together\n",
    "            \n",
    "            # Convert DNA to RNA (T -> U)\n",
    "            rna_sequence = sequence.replace(\"T\", \"U\").replace(\"t\", \"u\")\n",
    "            mrna_sequences[header] = rna_sequence\n",
    "            \n",
    "            print(f\"ðŸ“ƒ FASTA Header: {fasta_lines[0]}\")\n",
    "        \n",
    "        return mrna_sequences  # Return dictionary {Accession ID: mRNA sequence}\n",
    "    \n",
    "    except Exception as e:\n",
    "        print(f\"ðŸš¨ Error: {e}\")\n",
    "        return {}\n",
    "\n",
    "\n",
    "species = \"Saccharomyces cerevisiae\"\n",
    "gene = \"ACT1\"\n",
    "mrna_seqs = get_mrna_sequence(species, gene)\n",
    "for acc, seq in mrna_seqs.items():\n",
    "    print(f\"{acc}: {seq}\")\n",
    "    print(len(seq))  # print the first 60 bases for brevity\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetching mRNA sequence for:\n",
      "Species: Saccharomyces cerevisiae, Gene: CYC1\n",
      "\n",
      "Using refined query:\n",
      "Saccharomyces cerevisiae[Organism] AND CYC1[Gene] AND refseq_mrna[Filter]\n",
      "\n",
      "Raw esearch response (refseq_mrna filter):\n",
      "{'Count': '0', 'RetMax': '0', 'RetStart': '0', 'IdList': [], 'TranslationSet': [{'From': 'Saccharomyces cerevisiae[Organism]', 'To': '\"Saccharomyces cerevisiae\"[Organism]'}], 'QueryTranslation': 'Saccharomyces cerevisiae[Organism] AND CYC1[Gene] AND refseq_mrna[Filter]', 'ErrorList': {'PhraseNotFound': ['refseq_mrna[Filter]'], 'FieldNotFound': []}, 'WarningList': {'OutputMessage': ['No items found.'], 'QuotedPhraseNotFound': [], 'PhraseIgnored': []}}\n",
      "\n",
      "No records found with refseq_mrna filter.\n",
      "Using fallback query:\n",
      "Saccharomyces cerevisiae[Organism] AND CYC1[Gene] AND mRNA[Filter]\n",
      "Error: Search Backend failed: \t\t\n",
      "\n",
      "No sequence retrieved.\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "Entrez.email = \"qamilmirza@berkeley.edu\"\n",
    "\n",
    "def get_mrna_sequence_with_debug(species_name, gene_name):\n",
    "    \"\"\"\n",
    "    Fetches the mRNA sequence for a given species and gene.\n",
    "    Prints out the raw responses from the Entrez queries for debugging.\n",
    "    \n",
    "    Parameters:\n",
    "      species_name (str): e.g., \"Homo sapiens\" or \"Saccharomyces cerevisiae\"\n",
    "      gene_name (str): e.g., \"INS\" or \"CYC1\"\n",
    "      \n",
    "    Returns:\n",
    "      A single mRNA sequence (string) with T's replaced by U's,\n",
    "      or an empty string if no sequence is found.\n",
    "    \"\"\"\n",
    "    try:\n",
    "        # --- First Query: Using the refined query with refseq_mrna filter ---\n",
    "        search_query = f\"{species_name}[Organism] AND {gene_name}[Gene] AND refseq_mrna[Filter]\"\n",
    "        print(\"Using refined query:\")\n",
    "        print(search_query)\n",
    "        \n",
    "        handle = Entrez.esearch(db=\"nucleotide\", term=search_query, retmax=5)\n",
    "        record = Entrez.read(handle)\n",
    "        handle.close()\n",
    "        \n",
    "        print(\"\\nRaw esearch response (refseq_mrna filter):\")\n",
    "        print(record)\n",
    "        \n",
    "        # If no records are found, fall back to a generic query using mRNA filter\n",
    "        if not record[\"IdList\"]:\n",
    "            search_query = f\"{species_name}[Organism] AND {gene_name}[Gene] AND mRNA[Filter]\"\n",
    "            print(\"\\nNo records found with refseq_mrna filter.\")\n",
    "            print(\"Using fallback query:\")\n",
    "            print(search_query)\n",
    "            \n",
    "            handle = Entrez.esearch(db=\"nucleotide\", term=search_query, retmax=5)\n",
    "            record = Entrez.read(handle)\n",
    "            handle.close()\n",
    "            \n",
    "            print(\"\\nRaw esearch response (fallback mRNA filter):\")\n",
    "            print(record)\n",
    "            \n",
    "            if not record[\"IdList\"]:\n",
    "                print(\"\\nNo mRNA records found with either query.\")\n",
    "                return \"\"\n",
    "        \n",
    "        # Select the first gene ID (assumed to be the canonical transcript)\n",
    "        gene_id = record[\"IdList\"][0]\n",
    "        print(\"\\nSelected Gene ID:\")\n",
    "        print(gene_id)\n",
    "        \n",
    "        # --- Fetch the FASTA sequence for the selected record ---\n",
    "        fetch_handle = Entrez.efetch(db=\"nucleotide\", id=gene_id, rettype=\"fasta\", retmode=\"text\")\n",
    "        fasta_data = fetch_handle.read()\n",
    "        fetch_handle.close()\n",
    "        \n",
    "        print(\"\\nRaw FASTA data retrieved:\")\n",
    "        print(fasta_data)\n",
    "        \n",
    "        # Process FASTA data: remove header and join sequence lines\n",
    "        fasta_lines = fasta_data.strip().split(\"\\n\")\n",
    "        sequence_only = ''.join(fasta_lines[1:])\n",
    "        mrna_sequence = sequence_only.replace(\" \", \"\").replace(\"T\", \"U\").replace(\"t\", \"u\")\n",
    "        \n",
    "        return mrna_sequence\n",
    "    except Exception as e:\n",
    "        print(\"Error:\", e)\n",
    "        return \"\"\n",
    "\n",
    "# Example usage in your Jupyter Notebook:\n",
    "species_name = \"Saccharomyces cerevisiae\"\n",
    "gene_name = \"CYC1\"\n",
    "\n",
    "print(\"Fetching mRNA sequence for:\")\n",
    "print(f\"Species: {species_name}, Gene: {gene_name}\\n\")\n",
    "\n",
    "mrna_sequence = get_mrna_sequence_with_debug(species_name, gene_name)\n",
    "\n",
    "if mrna_sequence:\n",
    "    print(\"\\nFinal Processed mRNA Sequence (first 100 nucleotides):\")\n",
    "    print(mrna_sequence[:100])\n",
    "    print(\"\\nTotal Sequence Length:\", len(mrna_sequence), \"nucleotides\")\n",
    "else:\n",
    "    print(\"No sequence retrieved.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez\n",
    "\n",
    "Entrez.email = \"qamilmirza@berkeley.edu\"\n",
    "\n",
    "def get_gene_names(species_name):\n",
    "    \"\"\"\n",
    "    Fetch gene names for a species from NCBI.\n",
    "\n",
    "    Args:\n",
    "        species_name (str): e.g., \"Homo sapiens\"\n",
    "\n",
    "    Returns:\n",
    "        list: Sorted list of gene names.\n",
    "    \"\"\"\n",
    "    try:\n",
    "        # Search the ncbi gene database for gene names pertaining to the species but only take the first 30 for brevity\n",
    "        search_query = f\"{species_name}[Organism]\"\n",
    "        handle = Entrez.esearch(db=\"gene\", term=search_query, retmax=30)\n",
    "        record = Entrez.read(handle)\n",
    "        handle.close()\n",
    "\n",
    "        print(record)\n",
    "\n",
    "        # If we cannot find anything, then just return empty list\n",
    "        if not record[\"IdList\"]:\n",
    "            return []\n",
    "\n",
    "        # otgerwise get the gene ids\n",
    "        gene_ids = record[\"IdList\"]\n",
    "\n",
    "        # Then collect all the gene names using the gene ids to query the database\n",
    "        gene_names = []\n",
    "        for gene_id in gene_ids:\n",
    "            summary_handle = Entrez.esummary(db=\"gene\", id=gene_id)\n",
    "            summary_record = Entrez.read(summary_handle)\n",
    "            print(summary_record)\n",
    "            summary_handle.close()\n",
    "            if \"DocumentSummarySet\" in summary_record:\n",
    "                doc_summary = summary_record[\"DocumentSummarySet\"][\"DocumentSummary\"]\n",
    "                if doc_summary:\n",
    "                    gene_names.append(doc_summary[0][\"Name\"])\n",
    "\n",
    "        # I used a set to prevent duplicates and then sorted the list\n",
    "        return sorted(set(gene_names))\n",
    "    except Exception as e:\n",
    "        print(\"Error:\", e)\n",
    "        return []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Count': '360066', 'RetMax': '30', 'RetStart': '0', 'IdList': ['139440227', '139440225', '139440223', '139440221', '139440219', '139440217', '139440215', '139432819', '139432817', '139432815', '139432813', '139432811', '139432809', '139432807', '139432805', '139432803', '139432801', '139427322', '139355254', '139355252', '139355250', '139355248', '139355243', '139355241', '139355239', '139355237', '139355235', '139355233', '139355231', '139355229'], 'TranslationSet': [{'From': 'Homo sapiens[Organism]', 'To': '\"Homo sapiens\"[Organism]'}], 'TranslationStack': [{'Term': '\"Homo sapiens\"[Organism]', 'Field': 'Organism', 'Count': '360066', 'Explode': 'Y'}, 'GROUP'], 'QueryTranslation': '\"Homo sapiens\"[Organism]'}\n",
      "{'DocumentSummarySet': DictElement({'DocumentSummary': [DictElement({'Name': 'NACC1-AS1', 'Description': 'NACC1 antisense RNA 1', 'Status': '0', 'CurrentID': '0', 'Chromosome': '19', 'GeneticSource': 'genomic', 'MapLocation': '19p13.13', 'OtherAliases': '', 'OtherDesignations': '', 'NomenclatureSymbol': 'NACC1-AS1', 'NomenclatureName': 'NACC1 antisense RNA 1', 'NomenclatureStatus': 'Official', 'Mim': [], 'GenomicInfo': [], 'GeneWeight': '0', 'Summary': '', 'ChrSort': '~~last', 'ChrStart': '999999999', 'Organism': {'ScientificName': 'Homo sapiens', 'CommonName': 'human', 'TaxID': '9606'}, 'LocationHist': []}, attributes={'uid': '139440227'})], 'DbBuild': 'Build250209-2200m.1'}, attributes={'status': 'OK'})}\n",
      "{'DocumentSummarySet': DictElement({'DocumentSummary': [DictElement({'Name': 'PPP2R1A-AS1', 'Description': 'PPP2R1A antisense RNA 1', 'Status': '0', 'CurrentID': '0', 'Chromosome': '19', 'GeneticSource': 'genomic', 'MapLocation': '19q13.41', 'OtherAliases': '', 'OtherDesignations': '', 'NomenclatureSymbol': 'PPP2R1A-AS1', 'NomenclatureName': 'PPP2R1A antisense RNA 1', 'NomenclatureStatus': 'Official', 'Mim': [], 'GenomicInfo': [], 'GeneWeight': '0', 'Summary': '', 'ChrSort': '~~last', 'ChrStart': '999999999', 'Organism': {'ScientificName': 'Homo sapiens', 'CommonName': 'human', 'TaxID': '9606'}, 'LocationHist': []}, attributes={'uid': '139440225'})], 'DbBuild': 'Build250209-2200m.1'}, attributes={'status': 'OK'})}\n",
      "{'DocumentSummarySet': DictElement({'DocumentSummary': [DictElement({'Name': 'MAP2K7-AS1', 'Description': 'MAP2K7 antisense RNA 1', 'Status': '0', 'CurrentID': '0', 'Chromosome': '19', 'GeneticSource': 'genomic', 'MapLocation': '19p13.2', 'OtherAliases': '', 'OtherDesignations': '', 'NomenclatureSymbol': 'MAP2K7-AS1', 'NomenclatureName': 'MAP2K7 antisense RNA 1', 'NomenclatureStatus': 'Official', 'Mim': [], 'GenomicInfo': [], 'GeneWeight': '0', 'Summary': '', 'ChrSort': '~~last', 'ChrStart': '999999999', 'Organism': {'ScientificName': 'Homo sapiens', 'CommonName': 'human', 'TaxID': '9606'}, 'LocationHist': []}, attributes={'uid': '139440223'})], 'DbBuild': 'Build250209-2200m.1'}, attributes={'status': 'OK'})}\n",
      "{'DocumentSummarySet': DictElement({'DocumentSummary': [DictElement({'Name': 'CFAP97D1-AS1', 'Description': 'CFAP97D1 antisense RNA 1', 'Status': '0', 'CurrentID': '0', 'Chromosome': '17', 'GeneticSource': 'genomic', 'MapLocation': '17q21.31', 'OtherAliases': '', 'OtherDesignations': '', 'NomenclatureSymbol': 'CFAP97D1-AS1', 'NomenclatureName': 'CFAP97D1 antisense RNA 1', 'NomenclatureStatus': 'Official', 'Mim': [], 'GenomicInfo': [], 'GeneWeight': '0', 'Summary': '', 'ChrSort': '~~last', 'ChrStart': '999999999', 'Organism': {'ScientificName': 'Homo sapiens', 'CommonName': 'human', 'TaxID': '9606'}, 'LocationHist': []}, attributes={'uid': '139440221'})], 'DbBuild': 'Build250209-2200m.1'}, attributes={'status': 'OK'})}\n",
      "{'DocumentSummarySet': DictElement({'DocumentSummary': [DictElement({'Name': 'RAB27B-AS1', 'Description': 'RAB27B antisense RNA 1', 'Status': '0', 'CurrentID': '0', 'Chromosome': '18', 'GeneticSource': 'genomic', 'MapLocation': '18q21.2', 'OtherAliases': '', 'OtherDesignations': '', 'NomenclatureSymbol': 'RAB27B-AS1', 'NomenclatureName': 'RAB27B antisense RNA 1', 'NomenclatureStatus': 'Official', 'Mim': [], 'GenomicInfo': [], 'GeneWeight': '0', 'Summary': '', 'ChrSort': '~~last', 'ChrStart': '999999999', 'Organism': {'ScientificName': 'Homo sapiens', 'CommonName': 'human', 'TaxID': '9606'}, 'LocationHist': []}, attributes={'uid': '139440219'})], 'DbBuild': 'Build250209-2200m.1'}, attributes={'status': 'OK'})}\n",
      "{'DocumentSummarySet': DictElement({'DocumentSummary': [DictElement({'Name': 'MRM3-AS1', 'Description': 'MRM3 antisense RNA 1', 'Status': '0', 'CurrentID': '0', 'Chromosome': '17', 'GeneticSource': 'genomic', 'MapLocation': '17p13.3', 'OtherAliases': '', 'OtherDesignations': '', 'NomenclatureSymbol': 'MRM3-AS1', 'NomenclatureName': 'MRM3 antisense RNA 1', 'NomenclatureStatus': 'Official', 'Mim': [], 'GenomicInfo': [], 'GeneWeight': '0', 'Summary': '', 'ChrSort': '~~last', 'ChrStart': '999999999', 'Organism': {'ScientificName': 'Homo sapiens', 'CommonName': 'human', 'TaxID': '9606'}, 'LocationHist': []}, attributes={'uid': '139440217'})], 'DbBuild': 'Build250209-2200m.1'}, attributes={'status': 'OK'})}\n",
      "{'DocumentSummarySet': DictElement({'DocumentSummary': [DictElement({'Name': 'CELF6-AS1', 'Description': 'CELF6 antisense RNA 1', 'Status': '0', 'CurrentID': '0', 'Chromosome': '15', 'GeneticSource': 'genomic', 'MapLocation': '15q23', 'OtherAliases': '', 'OtherDesignations': '', 'NomenclatureSymbol': 'CELF6-AS1', 'NomenclatureName': 'CELF6 antisense RNA 1', 'NomenclatureStatus': 'Official', 'Mim': [], 'GenomicInfo': [], 'GeneWeight': '0', 'Summary': '', 'ChrSort': '~~last', 'ChrStart': '999999999', 'Organism': {'ScientificName': 'Homo sapiens', 'CommonName': 'human', 'TaxID': '9606'}, 'LocationHist': []}, attributes={'uid': '139440215'})], 'DbBuild': 'Build250209-2200m.1'}, attributes={'status': 'OK'})}\n",
      "{'DocumentSummarySet': DictElement({'DocumentSummary': [DictElement({'Name': 'PSMD3-AS1', 'Description': 'PSMD3 antisense RNA 1', 'Status': '0', 'CurrentID': '0', 'Chromosome': '', 'GeneticSource': 'genomic', 'MapLocation': '', 'OtherAliases': '', 'OtherDesignations': 'PR11-387H17.6|RP11-387H17.6', 'NomenclatureSymbol': 'PSMD3-AS1', 'NomenclatureName': 'PSMD3 antisense RNA 1', 'NomenclatureStatus': 'Official', 'Mim': [], 'GenomicInfo': [], 'GeneWeight': '164', 'Summary': '', 'ChrSort': '~~last', 'ChrStart': '999999999', 'Organism': {'ScientificName': 'Homo sapiens', 'CommonName': 'human', 'TaxID': '9606'}, 'LocationHist': []}, attributes={'uid': '139432819'})], 'DbBuild': 'Build250209-2200m.1'}, attributes={'status': 'OK'})}\n",
      "{'DocumentSummarySet': DictElement({'DocumentSummary': [DictElement({'Name': 'KRT84-AS1', 'Description': 'KRT84 and KRT82 antisense RNA 1', 'Status': '0', 'CurrentID': '0', 'Chromosome': '12', 'GeneticSource': 'genomic', 'MapLocation': '12q13.13', 'OtherAliases': '', 'OtherDesignations': '', 'NomenclatureSymbol': 'KRT84-AS1', 'NomenclatureName': 'KRT84 and KRT82 antisense RNA 1', 'NomenclatureStatus': 'Official', 'Mim': [], 'GenomicInfo': [], 'GeneWeight': '103', 'Summary': '', 'ChrSort': '~~last', 'ChrStart': '999999999', 'Organism': {'ScientificName': 'Homo sapiens', 'CommonName': 'human', 'TaxID': '9606'}, 'LocationHist': []}, attributes={'uid': '139432817'})], 'DbBuild': 'Build250209-2200m.1'}, attributes={'status': 'OK'})}\n",
      "{'DocumentSummarySet': DictElement({'DocumentSummary': [DictElement({'Name': 'KCNA6-AS1', 'Description': 'KCNA6 antisense RNA 1', 'Status': '0', 'CurrentID': '0', 'Chromosome': '12', 'GeneticSource': 'genomic', 'MapLocation': '12p13.32', 'OtherAliases': '', 'OtherDesignations': '', 'NomenclatureSymbol': 'KCNA6-AS1', 'NomenclatureName': 'KCNA6 antisense RNA 1', 'NomenclatureStatus': 'Official', 'Mim': [], 'GenomicInfo': [], 'GeneWeight': '103', 'Summary': '', 'ChrSort': '~~last', 'ChrStart': '999999999', 'Organism': {'ScientificName': 'Homo sapiens', 'CommonName': 'human', 'TaxID': '9606'}, 'LocationHist': []}, attributes={'uid': '139432815'})], 'DbBuild': 'Build250209-2200m.1'}, attributes={'status': 'OK'})}\n",
      "{'DocumentSummarySet': DictElement({'DocumentSummary': [DictElement({'Name': 'CEP152-AS1', 'Description': 'CEP152 antisense RNA 1', 'Status': '0', 'CurrentID': '0', 'Chromosome': '15', 'GeneticSource': 'genomic', 'MapLocation': '15q21.1', 'OtherAliases': '', 'OtherDesignations': '', 'NomenclatureSymbol': 'CEP152-AS1', 'NomenclatureName': 'CEP152 antisense RNA 1', 'NomenclatureStatus': 'Official', 'Mim': [], 'GenomicInfo': [], 'GeneWeight': '103', 'Summary': '', 'ChrSort': '~~last', 'ChrStart': '999999999', 'Organism': {'ScientificName': 'Homo sapiens', 'CommonName': 'human', 'TaxID': '9606'}, 'LocationHist': []}, attributes={'uid': '139432813'})], 'DbBuild': 'Build250209-2200m.1'}, attributes={'status': 'OK'})}\n",
      "{'DocumentSummarySet': DictElement({'DocumentSummary': [DictElement({'Name': 'OS9-AS1', 'Description': 'OS9 antisense RNA 1', 'Status': '0', 'CurrentID': '0', 'Chromosome': '12', 'GeneticSource': 'genomic', 'MapLocation': '12q13.3-q14.1', 'OtherAliases': '', 'OtherDesignations': '', 'NomenclatureSymbol': 'OS9-AS1', 'NomenclatureName': 'OS9 antisense RNA 1', 'NomenclatureStatus': 'Official', 'Mim': [], 'GenomicInfo': [], 'GeneWeight': '103', 'Summary': '', 'ChrSort': '~~last', 'ChrStart': '999999999', 'Organism': {'ScientificName': 'Homo sapiens', 'CommonName': 'human', 'TaxID': '9606'}, 'LocationHist': []}, attributes={'uid': '139432811'})], 'DbBuild': 'Build250209-2200m.1'}, attributes={'status': 'OK'})}\n",
      "{'DocumentSummarySet': DictElement({'DocumentSummary': [DictElement({'Name': 'CERS2-AS1', 'Description': 'CERS2 antisense RNA 1', 'Status': '0', 'CurrentID': '0', 'Chromosome': '1', 'GeneticSource': 'genomic', 'MapLocation': '1q21.3', 'OtherAliases': '', 'OtherDesignations': '', 'NomenclatureSymbol': 'CERS2-AS1', 'NomenclatureName': 'CERS2 antisense RNA 1', 'NomenclatureStatus': 'Official', 'Mim': [], 'GenomicInfo': [], 'GeneWeight': '103', 'Summary': '', 'ChrSort': '~~last', 'ChrStart': '999999999', 'Organism': {'ScientificName': 'Homo sapiens', 'CommonName': 'human', 'TaxID': '9606'}, 'LocationHist': []}, attributes={'uid': '139432809'})], 'DbBuild': 'Build250209-2200m.1'}, attributes={'status': 'OK'})}\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[20], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m names \u001b[38;5;241m=\u001b[39m \u001b[43mget_gene_names\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mHomo sapiens\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[16], line 34\u001b[0m, in \u001b[0;36mget_gene_names\u001b[0;34m(species_name)\u001b[0m\n\u001b[1;32m     32\u001b[0m gene_names \u001b[38;5;241m=\u001b[39m []\n\u001b[1;32m     33\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m gene_id \u001b[38;5;129;01min\u001b[39;00m gene_ids:\n\u001b[0;32m---> 34\u001b[0m     summary_handle \u001b[38;5;241m=\u001b[39m \u001b[43mEntrez\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mesummary\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdb\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mgene\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mid\u001b[39;49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgene_id\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     35\u001b[0m     summary_record \u001b[38;5;241m=\u001b[39m Entrez\u001b[38;5;241m.\u001b[39mread(summary_handle)\n\u001b[1;32m     36\u001b[0m     \u001b[38;5;28mprint\u001b[39m(summary_record)\n",
      "File \u001b[0;32m~/.pyenv/versions/3.11.9/lib/python3.11/site-packages/Bio/Entrez/__init__.py:345\u001b[0m, in \u001b[0;36mesummary\u001b[0;34m(**keywds)\u001b[0m\n\u001b[1;32m    343\u001b[0m variables\u001b[38;5;241m.\u001b[39mupdate(keywds)\n\u001b[1;32m    344\u001b[0m request \u001b[38;5;241m=\u001b[39m _build_request(cgi, variables)\n\u001b[0;32m--> 345\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_open\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.pyenv/versions/3.11.9/lib/python3.11/site-packages/Bio/Entrez/__init__.py:634\u001b[0m, in \u001b[0;36m_open\u001b[0;34m(request)\u001b[0m\n\u001b[1;32m    632\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(max_tries):\n\u001b[1;32m    633\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 634\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[43murlopen\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    635\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m HTTPError \u001b[38;5;28;01mas\u001b[39;00m exception:\n\u001b[1;32m    636\u001b[0m         \u001b[38;5;66;03m# Reraise if the final try fails\u001b[39;00m\n\u001b[1;32m    637\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m i \u001b[38;5;241m>\u001b[39m\u001b[38;5;241m=\u001b[39m max_tries \u001b[38;5;241m-\u001b[39m \u001b[38;5;241m1\u001b[39m:\n",
      "File \u001b[0;32m~/.pyenv/versions/3.11.9/lib/python3.11/urllib/request.py:216\u001b[0m, in \u001b[0;36murlopen\u001b[0;34m(url, data, timeout, cafile, capath, cadefault, context)\u001b[0m\n\u001b[1;32m    214\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    215\u001b[0m     opener \u001b[38;5;241m=\u001b[39m _opener\n\u001b[0;32m--> 216\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mopener\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mopen\u001b[49m\u001b[43m(\u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.pyenv/versions/3.11.9/lib/python3.11/urllib/request.py:519\u001b[0m, in \u001b[0;36mOpenerDirector.open\u001b[0;34m(self, fullurl, data, timeout)\u001b[0m\n\u001b[1;32m    516\u001b[0m     req \u001b[38;5;241m=\u001b[39m meth(req)\n\u001b[1;32m    518\u001b[0m sys\u001b[38;5;241m.\u001b[39maudit(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124murllib.Request\u001b[39m\u001b[38;5;124m'\u001b[39m, req\u001b[38;5;241m.\u001b[39mfull_url, req\u001b[38;5;241m.\u001b[39mdata, req\u001b[38;5;241m.\u001b[39mheaders, req\u001b[38;5;241m.\u001b[39mget_method())\n\u001b[0;32m--> 519\u001b[0m response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_open\u001b[49m\u001b[43m(\u001b[49m\u001b[43mreq\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdata\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    521\u001b[0m \u001b[38;5;66;03m# post-process response\u001b[39;00m\n\u001b[1;32m    522\u001b[0m meth_name \u001b[38;5;241m=\u001b[39m protocol\u001b[38;5;241m+\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_response\u001b[39m\u001b[38;5;124m\"\u001b[39m\n",
      "File \u001b[0;32m~/.pyenv/versions/3.11.9/lib/python3.11/urllib/request.py:536\u001b[0m, in \u001b[0;36mOpenerDirector._open\u001b[0;34m(self, req, data)\u001b[0m\n\u001b[1;32m    533\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m result\n\u001b[1;32m    535\u001b[0m protocol \u001b[38;5;241m=\u001b[39m req\u001b[38;5;241m.\u001b[39mtype\n\u001b[0;32m--> 536\u001b[0m result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_call_chain\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mhandle_open\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mprotocol\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mprotocol\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\n\u001b[1;32m    537\u001b[0m \u001b[43m                          \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m_open\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mreq\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    538\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m result:\n\u001b[1;32m    539\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "File \u001b[0;32m~/.pyenv/versions/3.11.9/lib/python3.11/urllib/request.py:496\u001b[0m, in \u001b[0;36mOpenerDirector._call_chain\u001b[0;34m(self, chain, kind, meth_name, *args)\u001b[0m\n\u001b[1;32m    494\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m handler \u001b[38;5;129;01min\u001b[39;00m handlers:\n\u001b[1;32m    495\u001b[0m     func \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mgetattr\u001b[39m(handler, meth_name)\n\u001b[0;32m--> 496\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    497\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m result \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    498\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "File \u001b[0;32m~/.pyenv/versions/3.11.9/lib/python3.11/urllib/request.py:1391\u001b[0m, in \u001b[0;36mHTTPSHandler.https_open\u001b[0;34m(self, req)\u001b[0m\n\u001b[1;32m   1390\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mhttps_open\u001b[39m(\u001b[38;5;28mself\u001b[39m, req):\n\u001b[0;32m-> 1391\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdo_open\u001b[49m\u001b[43m(\u001b[49m\u001b[43mhttp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mclient\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mHTTPSConnection\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mreq\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1392\u001b[0m \u001b[43m        \u001b[49m\u001b[43mcontext\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_context\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcheck_hostname\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_check_hostname\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.pyenv/versions/3.11.9/lib/python3.11/urllib/request.py:1348\u001b[0m, in \u001b[0;36mAbstractHTTPHandler.do_open\u001b[0;34m(self, http_class, req, **http_conn_args)\u001b[0m\n\u001b[1;32m   1346\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   1347\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 1348\u001b[0m         \u001b[43mh\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[43mreq\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_method\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mreq\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mselector\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mreq\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mheaders\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1349\u001b[0m \u001b[43m                  \u001b[49m\u001b[43mencode_chunked\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mreq\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mhas_header\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mTransfer-encoding\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1350\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mOSError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err: \u001b[38;5;66;03m# timeout error\u001b[39;00m\n\u001b[1;32m   1351\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m URLError(err)\n",
      "File \u001b[0;32m~/.pyenv/versions/3.11.9/lib/python3.11/http/client.py:1303\u001b[0m, in \u001b[0;36mHTTPConnection.request\u001b[0;34m(self, method, url, body, headers, encode_chunked)\u001b[0m\n\u001b[1;32m   1300\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mrequest\u001b[39m(\u001b[38;5;28mself\u001b[39m, method, url, body\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, headers\u001b[38;5;241m=\u001b[39m{}, \u001b[38;5;241m*\u001b[39m,\n\u001b[1;32m   1301\u001b[0m             encode_chunked\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m):\n\u001b[1;32m   1302\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Send a complete request to the server.\"\"\"\u001b[39;00m\n\u001b[0;32m-> 1303\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_send_request\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbody\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mheaders\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mencode_chunked\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.pyenv/versions/3.11.9/lib/python3.11/http/client.py:1349\u001b[0m, in \u001b[0;36mHTTPConnection._send_request\u001b[0;34m(self, method, url, body, headers, encode_chunked)\u001b[0m\n\u001b[1;32m   1345\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(body, \u001b[38;5;28mstr\u001b[39m):\n\u001b[1;32m   1346\u001b[0m     \u001b[38;5;66;03m# RFC 2616 Section 3.7.1 says that text default has a\u001b[39;00m\n\u001b[1;32m   1347\u001b[0m     \u001b[38;5;66;03m# default charset of iso-8859-1.\u001b[39;00m\n\u001b[1;32m   1348\u001b[0m     body \u001b[38;5;241m=\u001b[39m _encode(body, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mbody\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m-> 1349\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mendheaders\u001b[49m\u001b[43m(\u001b[49m\u001b[43mbody\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mencode_chunked\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mencode_chunked\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.pyenv/versions/3.11.9/lib/python3.11/http/client.py:1298\u001b[0m, in \u001b[0;36mHTTPConnection.endheaders\u001b[0;34m(self, message_body, encode_chunked)\u001b[0m\n\u001b[1;32m   1296\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1297\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m CannotSendHeader()\n\u001b[0;32m-> 1298\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_send_output\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmessage_body\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mencode_chunked\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mencode_chunked\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.pyenv/versions/3.11.9/lib/python3.11/http/client.py:1058\u001b[0m, in \u001b[0;36mHTTPConnection._send_output\u001b[0;34m(self, message_body, encode_chunked)\u001b[0m\n\u001b[1;32m   1056\u001b[0m msg \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;130;01m\\r\u001b[39;00m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mjoin(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_buffer)\n\u001b[1;32m   1057\u001b[0m \u001b[38;5;28;01mdel\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_buffer[:]\n\u001b[0;32m-> 1058\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmsg\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1060\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m message_body \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   1061\u001b[0m \n\u001b[1;32m   1062\u001b[0m     \u001b[38;5;66;03m# create a consistent interface to message_body\u001b[39;00m\n\u001b[1;32m   1063\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(message_body, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mread\u001b[39m\u001b[38;5;124m'\u001b[39m):\n\u001b[1;32m   1064\u001b[0m         \u001b[38;5;66;03m# Let file-like take precedence over byte-like.  This\u001b[39;00m\n\u001b[1;32m   1065\u001b[0m         \u001b[38;5;66;03m# is needed to allow the current position of mmap'ed\u001b[39;00m\n\u001b[1;32m   1066\u001b[0m         \u001b[38;5;66;03m# files to be taken into account.\u001b[39;00m\n",
      "File \u001b[0;32m~/.pyenv/versions/3.11.9/lib/python3.11/http/client.py:996\u001b[0m, in \u001b[0;36mHTTPConnection.send\u001b[0;34m(self, data)\u001b[0m\n\u001b[1;32m    994\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msock \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    995\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mauto_open:\n\u001b[0;32m--> 996\u001b[0m         \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconnect\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    997\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    998\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m NotConnected()\n",
      "File \u001b[0;32m~/.pyenv/versions/3.11.9/lib/python3.11/http/client.py:1468\u001b[0m, in \u001b[0;36mHTTPSConnection.connect\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1465\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mconnect\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[1;32m   1466\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mConnect to a host on a given (SSL) port.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m-> 1468\u001b[0m     \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconnect\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1470\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_tunnel_host:\n\u001b[1;32m   1471\u001b[0m         server_hostname \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_tunnel_host\n",
      "File \u001b[0;32m~/.pyenv/versions/3.11.9/lib/python3.11/http/client.py:962\u001b[0m, in \u001b[0;36mHTTPConnection.connect\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    960\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"Connect to the host and port specified in __init__.\"\"\"\u001b[39;00m\n\u001b[1;32m    961\u001b[0m sys\u001b[38;5;241m.\u001b[39maudit(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhttp.client.connect\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28mself\u001b[39m, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhost, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mport)\n\u001b[0;32m--> 962\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msock \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_create_connection\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    963\u001b[0m \u001b[43m    \u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mhost\u001b[49m\u001b[43m,\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mport\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msource_address\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    964\u001b[0m \u001b[38;5;66;03m# Might fail in OSs that don't implement TCP_NODELAY\u001b[39;00m\n\u001b[1;32m    965\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n",
      "File \u001b[0;32m~/.pyenv/versions/3.11.9/lib/python3.11/socket.py:836\u001b[0m, in \u001b[0;36mcreate_connection\u001b[0;34m(address, timeout, source_address, all_errors)\u001b[0m\n\u001b[1;32m    834\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m source_address:\n\u001b[1;32m    835\u001b[0m     sock\u001b[38;5;241m.\u001b[39mbind(source_address)\n\u001b[0;32m--> 836\u001b[0m \u001b[43msock\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconnect\u001b[49m\u001b[43m(\u001b[49m\u001b[43msa\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    837\u001b[0m \u001b[38;5;66;03m# Break explicitly a reference cycle\u001b[39;00m\n\u001b[1;32m    838\u001b[0m exceptions\u001b[38;5;241m.\u001b[39mclear()\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "names = get_gene_names(\"Homo sapiens\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "3.11.9",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
